students/anbaopham/Lesson_4/Assignment_2.py

```diff
@@ -0,0 +1,177 @@
+    with open(filename, "w") as fp:
+        fp.write(template.format(**letter_data))
+
+def get_first_last_name():
+    name_list = get_name_list()
+    fl_name_list =[]
+    for i in range(len(name_list)):
+        l = name_list[i].split()
+        if len(l)>2:
+            l[1:] =[" ".join(l[1:])]
+        fl_name_list.append(l)
+    return fl_name_list
+
+def Thank_You_letter_template(first_name, last_name, donation_amount):
+    first_name = first_name
+    last_name = last_name
+    donation_amount = donation_amount
+
+    template = """Dear {first_name} {last_name},
+    Thank you for your very kind donation of ${donation_amount}
+    It will be put to very good use.
+        Sincerely,
+            -The Team"""
+
+    letter_data ={"first_name": first_name, "last_name": last_name, "donation_amount": donation_amount}
+
+    filename = f"{first_name}_{last_name}.txt"
+    with open(filename, "w") as fp:
+        fp.write(template.format(**letter_data))
+
+
+
+def send_ALL_Thank_You():
+    fl_name_list = get_first_last_name()
+    FullName_list = get_name_list()
+    for i in range(len(fl_name_list)):
+        first_name = fl_name_list[i][0]
+        last_name = fl_name_list[i][1]
+        Thank_You_letter_template(first_name, last_name, 10)
+
+
+
+def create_a_Report():
+    l = (["Donor Name", "Total Given", "Num Gifts", "Average Gift"])
+    temp_list = []
+    l1 =[]
+
+    totalGiven =0
+    for x in donor_db:
+        temp_list.append(x[0])
+        tg = sum(list((x[1])))
+        totalGiven = str(tg)
+        temp_list.append(totalGiven)
+        ng = len(x[1])
+        numGift = str(ng)
+        temp_list.append(numGift)
+        #aveGift = str(round(totalGiven/numGift,2))
+        aG = round(tg/ng,2)
+        aveGift = str(aG)
+        temp_list.append(aveGift)
+        #report_table.append(x[0], totalGiven, numGift, aveGift)
+    m = 0
+    while m< len(temp_list):
+        l1.append(temp_list[m:m+4])
+        m = m+4
+
+    l1 =sorted(l1, key = sort_key)
+
+    l1.insert(0,l)
+
+    newL = []
+    for i in l1:
+        for j in i:
+            newL.append(len(j))
+            l = max(newL)
+    space = " "*10
+    width = l
+    #print(l1)
+    for a,b,c,d in l1:
+        print(f'{a:<{width}}{space}{b:>{width}}{space}{c:>{width}}{space}{d:>{width}}')
+
+def exit_program():
+    print("Bye!")
+    sys.exit()  # exit the interactive script
+
+response_dict ={"1": send_a_Thank_You, "2": create_a_Report, "3": send_ALL_Thank_You, "4": exit_program}
+def main():
+    while True:
+        response = input(prompt)  # continuously collect user selection
+        # now redirect to feature functions based on the user selection
+        if response in response_dict:
+            response_dict.get(response)()
+        else:
+            print("Not a valid option!")
+
+
+
+
+
+
+if __name__ == "__main__":
+    # don't forget this block to guard against your code running automatically if this module is imported
+    main()
```

Comment:
Same issues as prior version of mailroom.  Please clean up spacing and make sure variables and function names are snake_case.  Report needs to be sorted by total given in descending order.  Also, getting an error when adding a new donor to this version.

Reply:
fixed